### DISCUSSION TITLE
Can this project be built and used on open source code alone?

### DISCUSSION THREAD

**Discussion by ssddanbrown** [Budibase/budibase#12377] [Help]
Hello :wave:,

I've just been playing with the codebase, and just want to confirm if this project can be built and used from the open source code alone? 

I'm assuming it is, since the project is advertising itself as an open source project, but when attempting to build I'm seeing issues trying to resolve `@budibase/backend-core` files, which led me to notice [some "pro" mentioning](https://github.com/Budibase/budibase/blob/aed232852387ebe70704258d47a210839aa13a98/scripts/build.js#L29-L30) in some parts of the app (so not sure if it's related to licensing or pro components).

Additionally, from just looking at the source, I'm seeing some references to [use of a `@budibase/pro` package](https://github.com/Budibase/budibase/blob/master/packages/server/src/api/controllers/public/users.ts#L12) while this does not appear to be part of this git repo AFAICT (seems to be an empty submodule)?

**Comment by almereyda** [comment]
Would you mean freely-licensed open source, in your understanding of the term?

There are many ways to put source out in the open, depending on the license that is chosen.

This must not neccessarily contradict with licencing only parts of an enterprise application freely, and keeping enterprise-oriented features under proprietary licencing, which is also known as the "open-core" model. GitLab is a good example here.

The license statement is a good example of introducing ambiguity here. In https://github.com/Budibase/budibase/blob/master/LICENSE it reads:

> Copyright 2019-2021, Budibase Ltd.
>
> Each Budibase package has its own license, please check the license file in each package.
>
> You can consider Budibase to be GPLv3 licensed overall.
>
> The apps that you build with Budibase do not package any GPLv3 licensed code, thus do not fall under those restrictions.

I bet "overall" doesn't count as formal legal speech in any legislation, and thus all individual components that make up a final build package have to be considered for *each build*. Signed containers with SBOM will allow for provenance and integrity of this kind of metadata.

<details><summary>SBOM</summary>
<p>

* [How to add a Software Bill of Materials (SBOM) to your containers with GitHub Actions](https://actuated.dev/blog/sbom-in-github-actions)
* [sbom-tool/docs/setting-up-github-actions.md at main · microsoft/sbom-tool](https://github.com/microsoft/sbom-tool/blob/main/docs/setting-up-github-actions.md)
* [Introducing self-service SBOMs - The GitHub Blog](https://github.blog/2023-03-28-introducing-self-service-sboms/)
* [Generating SBOMs for Your Image with BuildKit | Docker](https://www.docker.com/blog/generate-sboms-with-buildkit/)
* [Generate the SBOM for Docker images | Docker Docs](https://docs.docker.com/engine/sbom/)

</p>
</details>

<details><summary>sigstore</summary>
<p>

* [Safeguard your containers with new container signing capability in GitHub Actions - The GitHub Blog](https://github.blog/2021-12-06-safeguard-container-signing-capability-actions/)

</p>
</details>

**Reply by pombredanne** [reply]
Actually GPL overall counts as a proper legal statement. This is the top level license asserted by the authors in the conventional license file. If anything there is something call promissory estoppel https://en.wikipedia.org/wiki/Estoppel and if this is NOT GPL, then this is a serious problem, as the top level GPL assertion would trump other lower level assertions made by the same authors, the key point being "overall" ...

If this is unclear this is a serious problem and there are other lingering license issues like:
- https://github.com/Budibase/budibase/issues/17514

Also the prebuilt proprietary SQS server binaries in https://github.com/Budibase/budibase/tree/master/hosting/couchdb/sqs/x86 make this conversation kind of moot if that code is required. 

This is also a potential (serious?) security issue as this is embeds NodeJS and OpenSSL and I would surmise that there could be vulnerabilities in NodeJS 18.20.4 and OpenSSL 3.0.13

**Comment by ssddanbrown** [comment]
> Would you mean freely-licensed open source, in your understanding of the term?

I mean open source as per the [commonly understood definition](https://opensource.org/osd/).

If an application is being openly advertised as being "open source", I'd expect to be able to utilise the core value advertised from open source code alone really. I'm not too fussed about the granular licensing setup used, or even there being versions that have non-open code, but If I cannot build and use the application from the sources shared as open source, then I'd consider advertising the project as a whole as open source quite misleading.

I'm not sure if that's the case here though, hence my original question.

**Reply by almereyda** [reply]
Good question. From what I can tell, the last release from

- https://github.com/Budibase/budibase/actions/workflows/tag-release.yml

was built in a non-public repository, so we cannot verify how and if the build works.

- https://github.com/Budibase/budibase/actions/runs/6825065010/job/18562187547#step:3:3

This points to a private repository `budibase/budibase-deploys`, where this trail of build provenance stops.

---

Did you check if you could run https://github.com/Budibase/budibase/blob/master/hosting/docker-compose.build.yaml?

Then it would be possible to run a manual build as described in https://github.com/Budibase/budibase/blob/master/docs/DEV-SETUP-DEBIAN.md, and provide feedback to their installation instructions there.

But you are right, of course, that this build procedure appears rather opaque, when considering the generic open source spirit of the GPLv3.

Regarding pro, the OSI doesn't consider it incompatible and their site explicitly states

> 4. Integrity of The Author’s Source Code
>
> The license may restrict source-code from being distributed in modified form only if the license allows the distribution of “patch files” with the source code for the purpose of modifying the program at build time. The license must explicitly permit distribution of software built from modified source code. The license may require derived works to carry a different name or version number from the original software.

The GPLv3 speaks about modified source and covered work, when they mean that *modified* versions are distributed. Yet here we see that the dependency is already part of the GPL'ed code, and if it would build sufficiently without it, I wouldn't see a violation. Yet.

The moment that the unreachable dependency is injected by a third-party and unknown means, a lawyer should be consulted.

But not if you are the owner of the copyright, because then you can give yourself different terms than the GPL dictates. This license is for us to use and/or modify/redistribute. As the IP belongs to Budibase Ltd., they can carry on with the code as they please.

My interpretation, but better ask a lawyer.

**Reply by ssddanbrown** [reply]
Thanks for the input!

> Did you check if you could run https://github.com/Budibase/budibase/blob/master/hosting/docker-compose.build.yaml?

Not yet, I was specifically following the [debian dev environment process](https://github.com/Budibase/budibase/blob/master/docs/DEV-SETUP-DEBIAN.md). TBH, I'm not actually running debian (running Fedora) but the dependencies looked platform abstract, and something I could meet.

> But not if you are the owner of the copyright, because then you can give yourself different terms than the GPL dictates.

Yeah, I can understand that Budibase can license how they wish, especially since they're getting contributors to hand over copyright via CLA, but my concern (if it's not usable on open source, still to be confirmed) is that it could be quite misleading by being advertised as an open source application. As an example, from the readme:

> Budibase is open-source - licensed as GPL v3. This should fill you with confidence that Budibase will always be around. You can also code against Budibase or fork it and make changes as you please, providing a developer-friendly experience. 

If I've bet on Budibase based on that, but later looked to fork for whatever reason, and found that chunks of the code were missing or proprietary, I'd be livid and feel misled. 

Again, still not sure if that's the case though, may just be build issues on my side, ideally waiting for some kind of official response just to confirm it's buildable/usable from the open source code before attempting further.

**Reply by alican92** [reply]
@ssddanbrown From what we can tell some parts of Budibase are proprietary and closed source. We've been trying to build BudiBase ourselves with our custom UI changes and we still run into the same issues.

If anything Budibase is becoming more closed source than open source.

The reasoning behind open source software was to allow developers unlimited freedom to fork and alter code and make desired changes. From what it seems Budibase has put limits even on the open source version. From requiring licenses to utilize the full functionality of the Public API to now imposing user limits on self hosted open source instances. We question how this can continue to be called opensource?

@joebudi Why all the limits and restrictions on Budibase and the software is still called Opensource?

**Reply by subos2008** [reply]
This brings up an interesting question: if a random person contributes to a GPLv3 project and attributes the copyright to the owner of the project. Can the assigned owner of the copyright then re-licence the contributed code?

That would seem rather contrary to the spirit of the GPL?

**Reply by ssddanbrown** [reply]
@subos2008 Quite often projects like this, under GPL based code, will [have a CLA](https://github.com/Budibase/budibase/blob/864fe6be6840b72fb66db29a7cbae0daf4c4e661/docs/CONTRIBUTING.md#contributor-license-agreement-cla) that [specifically hands over copyright](https://github.com/Budibase/budibase/blob/c8a12a66323341d120425b0ff7772506b17e3f9c/.github/cla/individual-cla.md?plain=1#L11) and allows re-licensing. For many folks, a CLA like this in a GPL project is a red flag, at least in respect to the commitment to open source, although this is very common since building upon and providing paid-for non-open variations is a common strategy (open core).
I think most would indeed see it as against the core spirit of the GPL.

**Comment by joshyorko** [comment]
So I've yet to find anybody who's actually built it from source. Hopefully some contributor or tryhard will point me in the right direction. I'm going to build it right now

**Comment by AI-Vectoring** [comment]
One of the downsides of Dockerization, no more installation guides...
By the way, this EE directory BS is catching on everywhere...
You can't call your project "Open Source" and then keep it half closed source, at the very least you need to have "separation of concerns"

**Reply by yudicon** [reply]
Absolutely agree. It's deceptive. Recently a German court decided that build scripts has also to be published alongside the code, because a user should be able to adopt the code so it's useful for him. (https://www.golem.de/news/klage-gegen-avm-auch-skripte-fuer-kompilierung-gehoeren-zum-quellcode-2501-192338.html)

**Comment by diraneyya** [comment]
I'm genuinely impressed with this phenomenal project, but I need clarity on its licensing structure. Does it follow the standard definition of open source? If it's open-core, which part constitutes the "core"? Is the community edition docker image completely open source?

The LICENSE file doesn't provide sufficient clarity. I'd suggest either:
- Splitting into two completely different repositories, or
- Following GitLab's approach with an ee folder and build flags to exclude premium features

To be clear, my concern isn't about paying for value—it's about understanding what functionality is available without payment, which helps justify subscription decisions.

The "user limit" raises questions about the open-core nature. If this is truly open source, what prevents modifying this limitation in the source code?

This appears to be a case of confused licensing rather than true open-core. From a place of genuine appreciation: please provide clearer boundaries so we can determine how much to rely on this outstanding software. As a consultant, I'm uncertain how I can leverage this project for client applications or revenue-generating prototypes.

Currently, this seems closer to n8n's ambiguous licensing approach rather than GitLab's clearer model.

**Comment by TRSx80** [comment]
Silence is deafening.

I'm just glad I found this while still in the initial research phase.

**Comment by ramnarayanank** [comment]
Yes, the Budibase repo can be built fully from the open-source code but you will notice references to @budibase/pro or pro modules. These are stubs for the enterprise features and are not required to run the community edition. When you build ensure dependencies point only to the OSS packages (e.g., @budibase/backend-core). The errors usually happen if submodules or workspace resolution are misconfigured. If you are aiming to [build a software product](https://www.contus.com/blog/how-to-build-a-software-product/) on top of Budibase, stick to the OSS repo. Pro packages are closed source and available separately.

**Reply by diraneyya** [reply]
Thank you for the response. I have been holding off using this project for any client work, while waiting for an answer. It is a pity, knowing that those who wrote this code want it being utilized and adopted by as many consultants as possible (please correct me if I am wrong there).

So the main question I have after hearing from you is whether the docker container budibase/budibase:latest contains the community (OSS) edition or includes pro features. More specifically:

- Does the latest Docker image enforce the 20-user limit that was introduced in late 2023, or does it provide unlimited users as true open-source software would?
- If there are user restrictions in the latest version, would using an older version like budibase/budibase:2.13.15 provide the unrestricted open-source experience?
- Can you clarify which specific features in the current Docker image require paid licensing versus what's available under the GPL v3 license?

I appreciate that businesses need sustainable revenue models, but as a consultant, I need to understand exactly what functionality I can rely on without ongoing licensing concerns for client deployments.

**Comment by pombredanne** [comment]
@melohagan can you clarify is this project is open source, GPL-licensed, or something else?

**Reply by melohagan** [reply]
Hey @pombredanne You can read more about our licensing here: https://github.com/Budibase/budibase?tab=License-1-ov-file#readme

*and within each package.

**Reply by pombredanne** [reply]
@melohagan I did read this before and it is confusing at best. For instance you state at https://github.com/Budibase/budibase/blob/7cdace0605a7563b56d570a73ff4b5e7ba1c37b9/LICENSE#L5 that "You can consider Budibase to be GPLv3 licensed overall" 

And then some code of yours is under an MPL license with an appendix that explicitly prohibits using a secondary license (e.g., restricts using the GPL)  https://github.com/Budibase/budibase/blob/7cdace0605a7563b56d570a73ff4b5e7ba1c37b9/packages/bbui/LICENSE#L371

These licenses are furthered modified which is likely problematic:
- https://github.com/Budibase/budibase/issues/17514


And there is proprietary SQS binary code with likely undisclosed/unattributed FOSS licenses, which may or may not be compatible with the GPL. See:
- https://github.com/Budibase/budibase/issues/17513


So, confused I am. :)

**Reply by joshyorko** [reply]
Because it's open source it's barely even source available. Community edition is absolute trash due to this fact. I have attempted about a year ago to even try and build it locally my removing the property bits but the whole thing is hardwired to not even build unless you have every single component attached including the paid edition stuff that in the community edition doesn't even work

**Reply by joshyorko** [reply]
It's not open source . Not even a little bit. It's smoke and mirrors repo to get you to sign up.   open source it's barely even source available. Community edition is absolute trash due to this fact. I have attempted about a year ago to  try and build it locally by removing the pro bits but the whole thing is hardwired to not even build unless you have every single component attached including the paid edition stuff that in the community edition doesn't even work. I'm not going to lie. I actually really do love the ux. The user experience the whole thing but companies like this are out to do absolutely one thing make money. They are not out here to help the community

**Reply by pombredanne** [reply]
@joshyorko Thanks for your input. One comment would suffice though ;)
